File: CLASS ACTIVITY/10.7.2026/s7_smartphone_market.py
import matplotlib.pyplot as plt
import seaborn as sns
import pandas as pd
import numpy as np

sns.set_style("whitegrid")
np.random.seed(1)

n = 50
brand = np.random.choice(['Samsung', 'Apple', 'OnePlus', 'Xiaomi'], n)
ram = np.random.choice([4, 6, 8, 12, 16], n)
storage = np.random.choice([64, 128, 256, 512], n)
speed = np.random.uniform(2.0, 3.5, n)
price = ram * 3000 + storage * 10 + speed * 5000 + np.random.normal(0, 3000, n)
df7 = pd.DataFrame({'Brand': brand, 'RAM': ram, 'Storage': storage, 'Speed': speed, 'Price': price})

grouped = df7.groupby(['RAM', 'Brand'], observed=True).agg(
    AvgPrice=('Price', 'mean'),
    AvgSpeed=('Speed', 'mean')
).reset_index()

plt.figure(figsize=(9, 6))
ax = sns.barplot(data=grouped, x='RAM', y='AvgPrice', hue='Brand', palette='viridis')

# annotate average processor speed on each bar (stands in for old color-gradient channel)
for container, brand_name in zip(ax.containers, grouped.Brand.unique()):
    sub_speed = grouped[grouped.Brand == brand_name]['AvgSpeed'].values
    ax.bar_label(container, labels=[f"{s:.1f}GHz" for s in sub_speed], fontsize=8, padding=2)

plt.title("RAM vs Avg Price by Brand (labels = Avg Processor Speed)")
plt.xlabel("RAM (GB)")
plt.ylabel("Average Price")
plt.tight_layout()
plt.show()

File: CLASS ACTIVITY/10.7.2026/RP/s7_smartphone_market.R
library(ggplot2)
library(dplyr)

set.seed(1)
n <- 50
brand <- sample(c('Samsung','Apple','OnePlus','Xiaomi'), n, replace=TRUE)
ram <- sample(c(4,6,8,12,16), n, replace=TRUE)
storage <- sample(c(64,128,256,512), n, replace=TRUE)
speed <- runif(n, 2.0, 3.5)
price <- ram*3000 + storage*10 + speed*5000 + rnorm(n, 0, 3000)
df7 <- data.frame(Brand=brand, RAM=ram, Storage=storage, Speed=speed, Price=price)

grouped <- df7 %>% group_by(RAM, Brand) %>%
  summarise(AvgPrice=mean(Price), AvgSpeed=mean(Speed), .groups='drop')

ggplot(grouped, aes(x=factor(RAM), y=AvgPrice, fill=Brand)) +
  geom_col(position=position_dodge()) +
  geom_text(aes(label=paste0(round(AvgSpeed,1),"GHz")),
            position=position_dodge(width=0.9), vjust=-0.3, size=3) +
  labs(title="RAM vs Avg Price by Brand (labels = Avg Processor Speed)",
       x="RAM (GB)", y="Average Price") +
  theme_minimal()
